### ISSUE TITLE
Default equalityComparer behavior

### ISSUE THREAD

**Issue by cervengoc** [knockout/knockout#2272]
Hi, is there any reason why the default `equalityComparer` doesn't support objects with referential equality? I mean I'd generally prefer an implementation similar to this.

```
var areEqual =
    arePrimitivesAndEqual(a,b) || 
    isObject(a) && !isArray(a) && a === b)
```

So it would basically add the support for `object`s excluding arrays. This would be useful for example in this case:

```
var viewModel = {
  items: ko.observableArray([ {id: 1, text: "1"},  {id: 2, text: "2"} ])
  selected: ko.observable(0),
  selectedItem: ko.computed(function() {
    var self = this;
    return this.items.filter(function(e) { return e.id === self.selected(); })[0];
  })
};
```

Here as far as I saw the `computed` will notify even if it evaluates to the same element.

**Comment by mbest** [comment]
Changing this default behavior would definitely be a breaking change. In your own application, you can configure this by changing the `ko.observable.fn.equalityComparer` and `ko.computed.fn.equalityComparer` properties.

**Comment by mbest** [comment]
You could also configure this on a per observable basis by creating an extender that sets this property for that observable. Perhaps that's also a direction we could take in Knockout, to expand the `notify` extender to support values in addition to `always`.

**Comment by cervengoc** [comment]
Would it really be a breaking change in practice? I see that it is in technical sense. But at first glance the only change would be that there were no redundant notifications in several cases.

**Comment by brianmhunt** [comment]
@cervengoc  I was wondering this too.

Certainly I'm ok with this being a breaking change in tko, and we can see in some environments if there's any negative impact.

**Comment by jrsearles** [comment]
It would certainly be a breaking change. Imagine the following:

```js
var obj = { foo: true };
var obs = ko.observable(obj);

// ... //

obj.foo = false;
obs(obj);
```

This would trigger now but not so if referential equality was used. I wouldn't say the above is a good practice but i would suspect there is a lot of code in the wild relying on the this behavior.

It's simple enough to override the equality operator as is, though it might be useful to have a few different built-in options. I have an extender i rely on for swapping out the comparer with a few different comparers depending on the purpose of the observable.

**Comment by cervengoc** [comment]
@jrsearles Yes it definitely is a breaking change however in the case of your example I wouldn't expect a notification. I think the more intuitive behavior is when there's no notification if the exact same instance is set. For these cases one can use `notify: 'always'` for example.
